[JSC] Implement `Array.prototype.flat` in C++

sosukesuzuki opened the pull request:
#### 7b9df2f4b1a9e2f12cf4feceb0870e44ab8efed6
<pre>
[JSC] Implement `Array.prototype.flat` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=294503">https://bugs.webkit.org/show_bug.cgi?id=294503</a>

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.flat` in C++.

The new fast path for `Array#flat` first calculates the length
of the resulting array using O(n) `calculateFlattenedLength`,
then directly sets values into the butterfly buffer of the
resulting array using O(n) `fastFlatIntoBuffer`.

                                             TipOfTree                  Patched

array-prototype-flat-depth-1-string      395.8826+-4.5561     ^    147.9479+-3.7062        ^ definitely 2.6758x faster
array-prototype-flat-large-nested        111.9910+-4.9277     ^     30.5805+-0.2314        ^ definitely 3.6622x faster
array-prototype-flat-sparse-array        198.6802+-2.0652     ^    133.4569+-3.7686        ^ definitely 1.4887x faster
array-prototype-flat-small-arrays          9.0792+-0.4023     ^      4.8345+-0.1067        ^ definitely 1.8780x faster
array-prototype-flat-depth-1-double      600.7011+-40.1338    ^    105.1259+-1.4612        ^ definitely 5.7141x faster
array-prototype-flat-depth-1-int32       447.6045+-6.5491     ^    106.2142+-7.6386        ^ definitely 4.2142x faster
array-prototype-flat-depth-infinity      524.9275+-4.1636     ^    141.2842+-5.6371        ^ definitely 3.7154x faster
array-prototype-flat-depth-2             522.6723+-2.1402     ^    118.4029+-5.5013        ^ definitely 4.4144x faster
array-prototype-flat-huge-arrays          43.4337+-3.9484     ^     10.1096+-0.1294        ^ definitely 4.2963x faster
array-prototype-flat-depth-1-mixed       450.7441+-18.1260    ^    136.2629+-0.7525        ^ definitely 3.3079x faster
array-prototype-flat-depth-3             526.4249+-2.3837     ^    124.7195+-4.6807        ^ definitely 4.2209x faster

* JSTests/microbenchmarks/array-prototype-flat-depth-1-double.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-int32.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-mixed.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-string.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-2.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-3.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-infinity.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-huge-arrays.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-large-nested.js: Added.
(shouldBe):
(createLargeNestedArray):
* JSTests/microbenchmarks/array-prototype-flat-small-arrays.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-sparse-array.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-flat-hole.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-flat-nested-proxy-array.js: Added.
(shouldBe):
(const.handler.has):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(flat): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::flatIntoArray):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::calculateFlattenedLength):
(JSC::fastFlatIntoBuffer):
(JSC::JSArray::fastFlat):
* Source/JavaScriptCore/runtime/JSArray.h:

Canonical link: <a href="https://commits.webkit.org/298127@main">https://commits.webkit.org/298127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2057d249afb9310b42173beeaa5c124157d7a23f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114347 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24556 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120519 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65065 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42653 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27652 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102694 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/67296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26833 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20819 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64201 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106746 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97027 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123729 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112910 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30847 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41738 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98894 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/95513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24338 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40655 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18491 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37403 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41241 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46749 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137119 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40836 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36681 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42584 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1023,6 +1023,178 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
+
+    switch (sourceType) {
+    case ArrayWithInt32:
+    case ArrayWithContiguous: {
```
We should have `ArrayWithInt32` separately since most of the code in this loop is only for `ArrayWithContiguous`

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -2175,6 +2176,105 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncToSpliced, (JSGlobalObject* globalObject,
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
```
We need to have recursion check via `vm.isSafeToRecurseSoft`. `flatIntoArray` is recursively calling, so it can be possible that we will have stackoverflow.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1023,6 +1023,178 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
+{
+    uint64_t resultLength = 0;
+    IndexingType sourceType = sourceArray->indexingType();
+
```
We need to have recursion check via `vm.isSafeToRecurseSoft`. `flatIntoArray` is recursively calling, so it can be possible that we will have stackoverflow.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1023,6 +1023,178 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_CATCH_SCOPE(vm);
+
+    IndexingType sourceType = sourceArray->indexingType();
```
We need to have recursion check via `vm.isSafeToRecurseSoft`. `flatIntoArray` is recursively calling, so it can be possible that we will have stackoverflow.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1040,6 +1040,199 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
+    switch (sourceType) {
+    case ArrayWithInt32:
+    case ArrayWithContiguous: {
+        auto* sourceBuffer = sourceArray->butterfly()->contiguous().data();
+        for (uint64_t i = 0; i < sourceLength; ++i) {
+            if (resultIndex >= vectorLength) [[unlikely]]
+                return std::numeric_limits<uint64_t>::max();
+            JSValue element = sourceBuffer[i].get();
+            if (!element) [[unlikely]]
+                continue;
+            if (depth > 0 && isJSArray(element)) {
+                JSArray* elementArray = jsCast<JSArray*>(element);
+                uint64_t newDepth = (depth == std::numeric_limits<uint64_t>::max()) ? depth : depth - 1;
+                resultIndex = fastFlatIntoBuffer(globalObject, resultBuffer, resultIndex, elementArray, elementArray->length(), newDepth, vectorLength);
+                RETURN_IF_EXCEPTION(scope, resultIndex);
+                if (resultIndex == std::numeric_limits<uint64_t>::max())
+                    return std::numeric_limits<uint64_t>::max();
+            } else {
+                if constexpr (std::is_same_v<T, double>)
+                    resultBuffer[resultIndex] = element.asNumber();
+                else
+                    resultBuffer[resultIndex].setWithoutWriteBarrier(element);
+                ++resultIndex;
+            }
+        }
```
Maybe we can have ArrayWithInt32 specific path since it never includes JSArray.

sosukesuzuki replied:
fixed!

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1040,6 +1040,199 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
     RELEASE_AND_RETURN(scope, joiner.join(globalObject));
 }
 
+static uint64_t calculateFlattenedLength(JSGlobalObject* globalObject, JSArray* sourceArray, uint64_t sourceLength, uint64_t depth)
```
I think there is a chance that we can craft a nested array which has more than uint64_t elements.

sosukesuzuki replied:
fixed with `CheckedUint64` !

Constellation reviewed (commented):
Left some comments.

sosukesuzuki committed:
[JSC] Implement `Array.prototype.flat` in C++
https://bugs.webkit.org/show_bug.cgi?id=294503

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.flat` in C++.

The new fast path for `Array#flat` first calculates the length
of the resulting array using O(n) `calculateFlattenedLength`,
then directly sets values into the butterfly buffer of the
resulting array using O(n) `fastFlatIntoBuffer`.

                                             TipOfTree                  Patched

array-prototype-flat-depth-1-string      395.8826+-4.5561     ^    147.9479+-3.7062        ^ definitely 2.6758x faster
array-prototype-flat-large-nested        111.9910+-4.9277     ^     30.5805+-0.2314        ^ definitely 3.6622x faster
array-prototype-flat-sparse-array        198.6802+-2.0652     ^    133.4569+-3.7686        ^ definitely 1.4887x faster
array-prototype-flat-small-arrays          9.0792+-0.4023     ^      4.8345+-0.1067        ^ definitely 1.8780x faster
array-prototype-flat-depth-1-double      600.7011+-40.1338    ^    105.1259+-1.4612        ^ definitely 5.7141x faster
array-prototype-flat-depth-1-int32       447.6045+-6.5491     ^    106.2142+-7.6386        ^ definitely 4.2142x faster
array-prototype-flat-depth-infinity      524.9275+-4.1636     ^    141.2842+-5.6371        ^ definitely 3.7154x faster
array-prototype-flat-depth-2             522.6723+-2.1402     ^    118.4029+-5.5013        ^ definitely 4.4144x faster
array-prototype-flat-huge-arrays          43.4337+-3.9484     ^     10.1096+-0.1294        ^ definitely 4.2963x faster
array-prototype-flat-depth-1-mixed       450.7441+-18.1260    ^    136.2629+-0.7525        ^ definitely 3.3079x faster
array-prototype-flat-depth-3             526.4249+-2.3837     ^    124.7195+-4.6807        ^ definitely 4.2209x faster

* JSTests/microbenchmarks/array-prototype-flat-depth-1-double.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-int32.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-mixed.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-1-string.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-2.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-3.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-depth-infinity.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-huge-arrays.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-large-nested.js: Added.
(shouldBe):
(createLargeNestedArray):
* JSTests/microbenchmarks/array-prototype-flat-small-arrays.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-prototype-flat-sparse-array.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-flat-hole.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-flat-nested-proxy-array.js: Added.
(shouldBe):
(const.handler.has):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(flat): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::flatIntoArray):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::calculateFlattenedLength):
(JSC::fastFlatIntoBuffer):
(JSC::JSArray::fastFlat):
* Source/JavaScriptCore/runtime/JSArray.h:

Canonical link: https://commits.webkit.org/298127@main
